### PR TITLE
[DM-10493] Add developer.lsst.io documentation for sqre-gtf.

### PR DESCRIPTION
**Pull request by jmatt**
Add documentation for sqre-gtf, Github branch protection and TravisCI.

	new file:   build-ci/github_protect_travis_flake8.rst
	modified:   index.rst

### CODE REVIEW

File: index.rst [hunk 1]
```diff
@@ -98,6 +98,7 @@ The `README <https://github.com/lsst-dm/dm_dev_guide/blob/master/README.md>`__ w
    build-ci/lsstsw.rst
    build-ci/new_package.rst
    build-ci/third_party.rst
+   build-ci/github_protect_travis_flake8.rst
```
Comment (timj) [inline index.rst]: `ci_overview.rst` was designed to be the root page for this. If @jonathansick decided that that page is not relevant we probably need to edit it to remove the Travis reference.

File: build-ci/github_protect_travis_flake8.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
+#############################################
+Github Branch Protection, TravisCI and Flake8
```
Comment (timj) [inline build-ci/github_protect_travis_flake8.rst]: There are distinct items here. Github configurations are not related to TravisCI jobs. I would expect "how to configure your github repo" to be in a different place in the developer guide. It should describe the recommended way of setting up a repository and provide the command line to automate that. Then you have a page describing what Travis can do and then informing the user how to enable specific options. Flake8 then becomes an example. Checking the build of latex/sphinx documents becomes another.

File: build-ci/github_protect_travis_flake8.rst [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+Flake8
+======
+
+`Flake8 <http://flake8.pycqa.org/en/stable/manpage.html>`_ is a command-line utility for enforcing style consistency across Python projects. It includes `PyFlakes <https://launchpad.net/pyflakes>`_ and `PEP-0008 <https://www.python.org/dev/peps/pep-0008/>`_ lint checks by default. DM uses the repository's :file:`setup.cfg` file to configure flake8. The :doc:`python style guide <../coding/python_style_guide#pep-8-is-the-baseline-coding-style>` has further details on specific lint check exceptions. The following :file:`setup.cfg` file implements DM's lint check exceptions:
+
+.. code-block:: text
+   [flake8]
+   max-line-length = 110
+   ignore = E133, E226, E228, N802, N803
```
Comment (timj) [inline build-ci/github_protect_travis_flake8.rst]: We should not duplicate the flake8 options here. Link to the version in the coding style. Note that `version.py` should be removed from the other copy as I fixed that recently.

File: build-ci/github_protect_travis_flake8.rst [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+.. _travisci:
+
+Travis CI
+=========
+
+TravisCI is a service that provides continuous integration for Github repositories. This service must be enabled for the repository before it will work. After it is enabled there will be a page available with the TravisCI status for the repository. For example lsst.utils `https://travis-ci.org/lsst/utils <https://travis-ci.org/lsst/utils>`_.
+
+.. _flake8:
+
+Flake8
```
Comment (timj) [inline build-ci/github_protect_travis_flake8.rst]: I believe @jdswinbank would like our Travis documentation to include descriptions of our Travis jobs. Out of scope for this ticket but we should design the layout of this document with that in mind (so the lsstsw checking, the latex/sphinx checking, flake8 checking).

File: build-ci/github_protect_travis_flake8.rst [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+           Type in your password to get an auth token from github
+           It will be stored in ~/.sq_github_token
+           and used in subsequent occasions.
+
+   Password for <your_github_login>: <your_github_password>
+   Enter 2FA code: <your_github_2FA>
+   Token written to ~/.sq_github_token
+
+There are three core features that sqre-gtf provides. Configuration of Github branch protection, enabling the TravisCI webhook and configuring your repository to use TravisCI to run Flake8 validation.
```
Comment (timj) [inline build-ci/github_protect_travis_flake8.rst]: font for `sqre-gtf`.

File: build-ci/github_protect_travis_flake8.rst [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+   [flake8]
+   max-line-length = 110
+   ignore = E133, E226, E228, N802, N803
+   exclude =
+       __init__.py,
+       .tests
+
+.. _sqre-gtf:
+
+sqre-gtf
```
Comment (timj) [inline build-ci/github_protect_travis_flake8.rst]: Opaque section title. Maybe "Automatically configuring your GitHub repository"?

File: build-ci/github_protect_travis_flake8.rst [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+========
+
+`sqre-gtf <https://github.com/lsst-sqre/sqre-gtf>`_ provides command line interfaces to use Github branch protection, TravisCI and Flake8. sqre-gtf runs on Python 2.7, 3.5 and 3.6. You can install it with
```
Comment (timj) [inline build-ci/github_protect_travis_flake8.rst]: `gtf` stands for what?

Comment (jmatt) [inline build-ci/github_protect_travis_flake8.rst]: Github TravisCI and Flake8 automation. But I am not attached to the name. I just needed to name the repository.

File: build-ci/github_protect_travis_flake8.rst (hunk 6 above)
Comment (jmatt) [inline build-ci/github_protect_travis_flake8.rst]: Yes this makes sense.

File: index.rst (hunk 1 above)
Comment (jmatt) [inline index.rst]: Yeah I saw this and didn't know what the right action was. I could use the ci_overview.rst or remove it. But it should be changed. @jonathansick is out sick today but I will discuss it with him when he's back.

File: build-ci/github_protect_travis_flake8.rst (hunk 5 above)
Comment (jmatt) [inline build-ci/github_protect_travis_flake8.rst]: I initially had it stylized. I looked around the dev guide and other repository references weren't necessarily stylize like they usually are in markdown rendered text. So I removed it. I'll add it back and bring this up with @jonathansick to verify. One example where it's definitely stylized is git-lfs, so I could do it similar to that.

**Comment by jonathansick** [comment]
Okay, sqre-gtf is a software product. We should keep the documentation for a software product with the product's code. Thus we should consider the README of https://github.com/lsst-sqre/sqre-gtf as the product's documentation, and focus on making that README as usable as possible (it's already got a good start). If desired, we should create `https://sqre-gtf.lsst.io` that is built from the content of the README.

As for the Developer Guide, yes, we need more guidelines on how we use CI in DM and how we configure GitHub repos in DM. But I think that what folks are asking for is way beyond the scope of this particular ticket.

At most I'd have a page in the Build/CI/Testing section that says "Setting up Flake8 Testing for Python Codebases" or the like, and all it says is something like:

> The DM Python Style Guide reflects PEP 8 with [some exceptions](https://developer.lsst.io/coding/python_style_guide.html#code-may-be-validated-with-flake8). To maintain conformity with the Style Guide, we recommend that all Python projects run flake8 as part of CI, as described in the DM Python Style Guide.
>
> sqre-gtf is a tool for configuring Travis CI to automatically run on Git pushes. Using GitHub's branch protections, the Travis checks that sqre-gtf enables prevent pull requests from being merged that have flake8 errors.
>
> For information on configuring a Python project to run Flake8 in Travis CI and enable GitHub branch protectors, see the [sqre-gtf documentation](https://github.com/lsst-sqre/sqre-gtf).

**Comment by ktlim** [comment]
Since some Travis and GitHub protection documentation has been added to the Dev Guide, is this PR now obsolete?

**Pull request closed without merging**